csrc/scheduler/pointwise_utils.cpp

```diff
@@ -142,6 +142,115 @@ bool DomainMap::areAllInputIdsMappedTo(TensorView* input_tv, TensorView* tv)
+// can't be reversed. The check in this function would give us T34 with source
+// i0, i1; where T198 would have source i0, b3, i1, where b3 isn't contained in
+// T34. Hence we'll reject this reference_tv.
+//
+// ------
+//
+// e.g 1.
+//   T0 [i0, i1]
+//   T1 [i2, i0, i1]
+//   T2 [i0*i1]      = reshape(T0)
+//   T3 [b3, i0, i1] = broadcast(T0)
+//   T4 [i2, i0, i1] = add(T1, T3)
+//   output(T2)
+//   output(T4)
+//
+// the example above should be able to pick T4 as reference_tv. T2's source i0,
+// i1 are both contained by the source of T4, so this example could be scheduled
+// as a single fusion.
+bool DomainMap::areAllOutputIdsMappedTo(
+    TensorView* output_tv,
+    TensorView* reference_tv) const {
+  // traverse back to collect all disjoint set producers from the logical domain
+  // of tv.
+  auto get_source_producers = [this](TensorView* tv) {
+    VectorOfUniqueEntries<std::shared_ptr<VectorOfUniqueEntries<IterDomain*>>>
+        all_producer_sets;
+    std::for_each(
+        tv->getLogicalDomain().begin(),
+        tv->getLogicalDomain().end(),
+        [&](IterDomain* id) {
+          all_producer_sets.pushBack(
+              ca_map_.disjointSetOf(id, IdMappingMode::EXACT));
+        });
+    all_producer_sets.pushBack(
+        ca_map_.getAllDisjointSetProducers(all_producer_sets));
+
+    std::vector<IterDomain*> source_ids;
+    std::for_each(
+        all_producer_sets.vector().begin(),
+        all_producer_sets.vector().end(),
+        [&source_ids,
+         this](const std::shared_ptr<VectorOfUniqueEntries<IterDomain*>>&
+                   producer_set_ptr) {
+          IterDomain* id = producer_set_ptr->front();
+          if (ca_map_.uniqueExactDefinitions(id).empty()) {
+            source_ids.push_back(id);
+          }
+        });
+    return source_ids;
+  };
+
+  // this contains all source iter domain that's covered by reference_tv, so
+  // it's safe for output_tv to have them.
+  std::unordered_set<IterDomain*> covered_source_ids;
+  for (IterDomain* id : get_source_producers(reference_tv)) {
+    covered_source_ids.insert(id);
+  }
+  // Note: there's certain cases where it's safe to have dangling IDs,
+  // e.g
+  //   T34  [i0, i1]
+  //   T185 [i0, b2, i1]     = broadcast(T34)
+  //   T192 [i0, b3(ex), i1] = expand(T185)
+  // It's safe to propagate T34 to T192, since b3(ex) is not involved in the
+  // propagation. But this isn't generally safe. If the above example is changed
+  // to e.g
+  //   T34  [i0, i1]
+  //   T185 [i0, b2, i1]     = broadcast(T34)
+  //   T186 [i0, i4, i1]     = ones({i0, i4, i1})
+  //   T193 [i0, i4, i1]     = add(T34, T186)
+  // It's unsafe to propagate from T34 to T193, see issue
+  // https://github.com/NVIDIA/Fuser/issues/3542
+
+  // Check all source iter domain involved in producing output_tv
+  for (IterDomain* id : get_source_producers(output_tv)) {
+    // It's safe to have unmapped broadcast dimension
+    if (id->isBroadcast()) {
```

Comment:
This is the change from last review. I think it's safe to ignore broadcast dimensions that's not mapped?

I also think this should be done in the old loop instead, as vvv
```
  std::unordered_set<IterDomain*> covered_source_ids;
  for (IterDomain* id : output_tv->getLogicalDomain()) {
    if (id->isBroadcast()) {
      covered_source_ids.insert(id);
    }
  }
```

Given that right now `expand` is not visible on ID in fusion IR, I don't think it makes a difference one way or the other.